src/web/middlewares/web_middleware_csrf.ml

```diff
@@ -1,6 +1,80 @@
-(* TODO
-https://docs.djangoproject.com/en/3.0/ref/csrf/#how-it-works *)
+open Lwt.Syntax
 
-let m () =
-  let filter handler ctx = handler ctx in
-  Web_middleware_core.create ~name:"csrf" filter
+let ctx_token_key : Token.t Core.Ctx.key = Core.Ctx.create_key ()
+
+(*TODO [aerben] optional*)
+let get_token ctx = Core_ctx.find ctx_token_key ctx
+
+exception No_csrf_token of string
+
+(* TODO [aerben] rebase *)
+(* TODO [aerben] salt token *)
+(* TODO [aerben] remove unit for non-optional fns *)
+(* TODO [aerben] use sessionservice, see if already has unexpired token, if yes, just create new salt, if no, create new token and new salt, maybe see if sessions expire too, then dont need token expiry, salting and hashing has to be fast! *)
+(* TODO (https://docs.djangoproject.com/en/3.0/ref/csrf/#how-it-works)
+Check other Django specifics namely:
+   - Testing views with custom HTTP client
+   - Allow Sihl user to make views exempt
+   - Enable subdomain
+   - HTML caching token handling
+*)
+module Make
+    (TokenService : Token.Sig.SERVICE)
+    (Log : Log.Sig.SERVICE)
+    (SessionService : Session.Sig.SERVICE) =
+struct
+  let m () =
+    let filter handler ctx =
+      (* Create a token no matter the HTTP request type *)
+      let* token = TokenService.create ctx ~kind:"csrf" () in
+      (* Store the ID in the session *)
+      (* Storing the token directly could mean it ends up on the client
+         if the cookie backend is used for session storage *)
+      let* () = SessionService.set_value ctx ~key:"csrf" ~value:token.id in
+      let ctx = Core_ctx.add ctx_token_key token ctx in
+      (* Don't check for CSRF token in GET requests *)
+      (* TODO don't check for HEAD, OPTIONS and TRACE either *)
+      if Web_req.is_get ctx then handler ctx
+      else
+        let* value = Web_req.urlencoded ctx "csrf" in
+        match value with
+        (* Give 403 if no token provided *)
+        | Error _ -> Web_res.(html |> set_status 403) |> Lwt.return
+        | Ok value -> (
+            let* id = SessionService.get_value ctx ~key:"csrf" in
+            let token_id =
+              match id with
+              | None ->
+                  (* TODO [aerben] SHOULD THIS BE 403? *)
```

Comment:
@jerben What do you think? This is when there is no CSRF token in a users session. This implies the user has not made a GET or cleared their cookies after loading a form (or maybe another scenario I can't think of now). Should we do 403 in this case?

Reply:
@aronerben I think it should be 403, yes.